src/recipe_dictionary.cpp

```diff
@@ -32,10 +32,121 @@ const recipe &recipe_dictionary::get_uncraft( const itype_id &id )
 std::vector<const recipe *> recipe_subset::search( const std::string &txt ) const
 {
     std::vector<const recipe *> res;
-
+    std::vector<std::string> subqueries;
+    std::vector<std::string> requirement_qualities,
+        names,
+        components,
+        tools,
+        result_qualities,
+        skills;
+    std::string pattern = txt; //Copy the search string.
+    size_t split;
+    if( ( split = pattern.find( ',' ) ) != std::string::npos ) {
+        while( ( split = pattern.find( ',' ) ) != std::string::npos ) {
+            subqueries.push_back( pattern.substr( 0, split ) );
+            pattern = pattern.substr( split + 1 );
+        }
+    } else {
+        subqueries.push_back( pattern );
+    }
+    for( const auto &query : subqueries ) {
+        if( ( split = query.find( ':' ) ) == std::string::npos ) { //search by name
+            names.push_back( query );
+        } else {
+            switch( query[split - 1] ) { //Find the appropriate key
+                case 'Q'://search by quality
+                    result_qualities.push_back( query.substr( split + 1 ) );
+                    break;
+                case 'q':
+                    requirement_qualities.push_back( query.substr( split + 1 ) );
+                    break;
+                case 'c'://search by component
+                    components.push_back( query.substr( split + 1 ) );
+                    break;
+                case 't'://search by requirement
+                    tools.push_back( query.substr( split + 1 ) );
+                    break;
+                case 's'://search by skill.
+                    skills.push_back( query.substr( split + 1 ) );
+                    break;
+            }
+        }
+    }
+    auto name_match =
+    [names]( const recipe * r ) {
+        return std::any_of( names.begin(), names.end(),
+        [r]( const std::string & name ) {
```

Comment:
Try for the nested lambda on one line maybe? It's quite hard to read this section